scalding-commons-macros/src/main/scala/com/twitter/scalding/commons/macros/impl/ordser/UnionLike.scala

```diff
@@ -0,0 +1,279 @@
+        val extract = optiTBuf.map { tBuf =>
+          q"""
+            ${tBuf.get(inputStream)}
+          """
+        }.getOrElse {
+          q"""(new Object).asInstanceOf[$tpe]"""
+        }
+
+        existing match {
+          case Some(t) =>
+            Some(q"""
+            if($valueA == $idx) {
+              $extract : $tpe
+            } else {
+              $t
+            }
+          """)
+          case None =>
+            Some(q"""
+          if($valueA == $idx) {
+            $extract
+          } else {
+            sys.error("Did not understand thrift union idx: " + $valueA)
+          }
+            """)
+        }
+    }.get
+
+    q"""
+        val $valueA: Int = $inputStream.readByte.toInt
+        $expandedOut
+      """
+  }
+
+  def compare(c: Context)(cmpType: c.Type, elementA: c.TermName, elementB: c.TermName)(subData: List[(Int, c.Type, Option[TreeOrderedBuf[c.type]])]): c.Tree = {
+    import c.universe._
+
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    val arg = freshT("arg")
+    val idxCmp = freshT("idxCmp")
+    val idxA = freshT("idxA")
+    val idxB = freshT("idxB")
+
+    val toIdOpt: Tree = subData.foldLeft(Option.empty[Tree]) {
+      case (existing, (idx, tpe, _)) =>
+        existing match {
+          case Some(t) =>
+            Some(q"""
+            if($arg.isInstanceOf[$tpe]) {
+              $idx
+            } else {
+              $t
+            }
+          """)
+          case None =>
+            Some(q"""
+              if($arg.isInstanceOf[$tpe]) {
+                $idx
+              } else {
+                sys.error("Unable to compare unknown type")
+              }""")
+        }
+    }.get
+
+    val compareSameTypes: Option[Tree] = subData.foldLeft(Option.empty[Tree]) {
+      case (existing, (idx, tpe, optiTBuf)) =>
+        val commonCmp = optiTBuf.map { tBuf =>
+          val aTerm = freshT("aTerm")
+          val bTerm = freshT("bTerm")
+          q"""
+          val $aTerm: $tpe = $elementA.asInstanceOf[$tpe]
+          val $bTerm: $tpe = $elementB.asInstanceOf[$tpe]
+          ${tBuf.compare(aTerm, bTerm)}
+        """
+        }.getOrElse(q"0")
+
+        existing match {
+          case Some(t) =>
+            Some(q"""
+            if($idxA == $idx) {
+              $commonCmp : Int
+            } else {
+              $t : Int
+            }
+          """)
+          case None =>
+            Some(q"""
+              if($idxA == $idx) {
+                $commonCmp : Int
+              } else {
+                $idxCmp
+              }""")
+        }
+    }
+
+    val compareFn = q"""
+      def instanceToIdx($arg: $cmpType): Int = {
+        ${toIdOpt}: Int
+      }
+
+      val $idxA: Int = instanceToIdx($elementA)
+      val $idxB: Int = instanceToIdx($elementB)
+      val $idxCmp: Int = $idxA.compare($idxB)
+
+      if($idxCmp != 0) {
+        $idxCmp: Int
+      } else {
+        ${compareSameTypes.get}: Int
```

Comment:
can you comment why you break the pattern here of not calling .get at the end of the fold and instead put it here? Is it that sometimes this would fail (in the other branches)?